src/v/security/authorizer.h

```diff
@@ -290,21 +292,111 @@ class authorizer final {
               bool(_allow_empty_matches));
         }
 
-        // check for deny
-        if (auto entry = acls.find(
-              operation, principal, host, acl_permission::deny);
-            entry.has_value()) {
-            return auth_result::acl_match(
-              principal, host, operation, resource_name, false, *entry);
+        auto check_access =
+          [this, &acls, &operation, &host, &resource_name](
+            const security::acl_principal& user,
+            std::optional<const security::acl_principal_base*> role,
+            acl_permission perm) -> std::optional<auth_result> {
+            assert(
+              !role
+              || *role != nullptr && (*role)->type() == principal_type::role);
+            const acl_principal_base& to_check = *role.value_or(&user);
+            bool is_allow = perm == acl_permission::allow;
+            std::optional<acl_matches::acl_match> entry;
+            if (is_allow) {
+                entry = acl_any_implied_ops_allowed(
+                  acls, to_check, host, operation);
+            } else {
+                entry = acls.find(operation, to_check, host, perm);
+            }
+            if (!entry) {
+                return std::nullopt;
+            }
+            switch (to_check.type()) {
+            case principal_type::user:
+            case principal_type::ephemeral_user:
+                return auth_result::acl_match(
+                  user, host, operation, resource_name, is_allow, *entry);
+            case principal_type::role:
+                return auth_result::role_acl_match(
+                  user,
+                  security::role_name{to_check.name_view()},
+                  host,
+                  operation,
+                  resource_name,
+                  is_allow,
+                  *entry);
+            }
+            __builtin_unreachable();
+        };
+
+        auto check_deny =
+          [&check_access](
+            const security::acl_principal& user,
+            std::optional<const security::acl_principal_view*> role
+            = std::nullopt) {
+              return check_access(user, role, acl_permission::deny);
+          };
+
+        auto check_allow =
+          [&check_access](
+            const security::acl_principal& user,
+            std::optional<const security::acl_principal_view*> role
+            = std::nullopt) {
+              return check_access(user, role, acl_permission::allow);
+          };
```

Comment:
These wrappers are syntactically quite noise to curry a single parameter, I think it might be simpler to pass `acl_permission` as the first parameter of `check_access` and `access_view`

Reply:
I'm cool with that

Reply:
done